adaptris/interlok
# INTERLOK-3561 ThreadContextWorkflow

## Motivation

From @gerco 

>When running a bunch of PoolingWorkflows with an embedded Jetty connection, I'm noticing various performance issues depending on configuration:
>
>If the jetty executor is set to some number threads and each pooling workflow is set to fewer threads, a situation can occur where one PoolingWorkflow is swamped with work and has dozens of Jetty threads waiting for an available thread in the PoolingWorkflow.
>
>If the Jetty executor is set to the some number of threads and each pooling workflow is set to the same number, we are running many more threads than we ideally would. This is causing issues with dynamic-shared-services caching since these caches can be very large and per-thread so we run out of memory.
>
>Is there a way to run workflows directly on the Jetty threads rather than have a PoolingWorkflow in between, introducing a layer of complexity we don't need?


## Modification

- Refactor a shared hierarchy out of PoolingWorkflow into a "WorkflowWithObjectPool"
- Added a new ThreadContextWorkflow that extends WorkflowWithObjectPool
    - Not synchronized and maintains a configurable object pool to avoid non-thread safe service lists.
- Modify JettyNoBacklogInterceptor to use WorkflowWithObjectPool as its decision point where it previously used PoolingWorkflow.

## Result

A new workflow `ThreadContextWorkflow` is now available to configure.

## Testing

- Have a jetty workflow that contains a standard workflow; the service should just be WaitService (default 20 seconds is fine, but you'll get bored) : `/api/standard-sleep`
- Have a jetty workflow that contains a pooling workflow; the service should be exactly the same as above : `/api/pooled-sleep`
- Test them both with 3 (or more)  simultaneous curl requests and check that behaviour is as expected.
    - the pooled version will all finish in ~20 seconds.
    - The standard version will finish in ~60 seconds (or more), since standardworkflow is synchronized.

- Create a new thread-context-workflow that has the same configuration : `/api/sleep`
    - additional-debug (expert mode) : true (for extra logging).
    - use-current-thread-name : true (this is the default).
- Test the new workflow using simultaneous curl requests...

- This will finish in ~20 seconds proving that we're handling the requests in a threaded fashion that is unrelated to any internal thread pools that we have.

You should have logging that looks something like this, the __qtp*__ is the internal jetty thread pool thread.

```
TRACE [sad-hugle@default-channel(qtp674484384-35)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Borrow]: Idle [0], Active [1], Max [10]
DEBUG [sad-hugle@default-channel(qtp674484384-35)] [c.a.c.ThreadContextWorkflow.handleMessage()] [{}] start processing msg...
DEBUG [sad-hugle@default-channel(qtp674484384-35)] [c.a.c.ServiceList.applyServices()] [{}] Executing doService on [WaitService(loving-dijkstra)]
TRACE [sad-hugle@default-channel(qtp674484384-37)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Borrow]: Idle [0], Active [2], Max [10]
DEBUG [sad-hugle@default-channel(qtp674484384-37)] [c.a.c.ThreadContextWorkflow.handleMessage()] [{}] start processing msg...
DEBUG [sad-hugle@default-channel(qtp674484384-37)] [c.a.c.ServiceList.applyServices()] [{}] Executing doService on [WaitService(loving-dijkstra)]
TRACE [sad-hugle@default-channel(qtp674484384-41)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Borrow]: Idle [0], Active [3], Max [10]
DEBUG [sad-hugle@default-channel(qtp674484384-41)] [c.a.c.ThreadContextWorkflow.handleMessage()] [{}] start processing msg...
DEBUG [sad-hugle@default-channel(qtp674484384-41)] [c.a.c.ServiceList.applyServices()] [{}] Executing doService on [WaitService(loving-dijkstra)]

...
TRACE [sad-hugle@default-channel(qtp674484384-35)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Return]: Idle [1], Active [2], Max [10]
TRACE [sad-hugle@default-channel(qtp674484384-37)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Return]: Idle [2], Active [1], Max [10]
TRACE [sad-hugle@default-channel(qtp674484384-41)] [c.a.c.ThreadContextWorkflow.logPoolState()] [{}] Current Pool after [Return]: Idle [3], Active [0], Max [10]
```
